beefy-gadget/src/lib.rs

```diff
@@ -52,74 +55,89 @@ pub fn beefy_peers_set_config() -> sc_network::config::NonDefaultSetConfig {
 	}
 }
 
+/// A convenience BEEFY client trait that defines all the type bounds a BEEFY client
+/// has to satisfy. Ideally that should actually be a trait alias. Unfortunately as
+/// of today, Rust does not allow a type alias to be used as a trait bound. Tracking
+/// issue is <https://github.com/rust-lang/rust/issues/41517>.
+pub(crate) trait Client<B, BE, P>:
+	BlockchainEvents<B> + HeaderBackend<B> + Finalizer<B, BE> + ProvideRuntimeApi<B> + Send + Sync
+where
+	B: Block,
+	BE: Backend<B>,
+	P: sp_core::Pair,
+	P::Public: AppPublic + Codec,
+	P::Signature: Clone + Codec + Debug + PartialEq + TryFrom<Vec<u8>>,
+{
+	// empty
+}
+
+impl<B, BE, P, T> Client<B, BE, P> for T
+where
+	B: Block,
+	BE: Backend<B>,
+	P: sp_core::Pair,
+	P::Public: AppPublic + Codec,
+	P::Signature: Clone + Codec + Debug + PartialEq + TryFrom<Vec<u8>>,
+	T: BlockchainEvents<B> + HeaderBackend<B> + Finalizer<B, BE> + ProvideRuntimeApi<B> + Send + Sync,
+{
+	// empty
+}
+
 /// Allows all gossip messages to get through.
 struct AllowAll<Hash> {
 	topic: Hash,
 }
 
-impl<Block> GossipValidator<Block> for AllowAll<Block::Hash>
+impl<B> GossipValidator<B> for AllowAll<B::Hash>
 where
-	Block: BlockT,
+	B: Block,
 {
 	fn validate(
 		&self,
-		_context: &mut dyn GossipValidatorContext<Block>,
+		_context: &mut dyn GossipValidatorContext<B>,
 		_sender: &sc_network::PeerId,
 		_data: &[u8],
-	) -> GossipValidationResult<Block::Hash> {
+	) -> GossipValidationResult<B::Hash> {
 		GossipValidationResult::ProcessAndKeep(self.topic)
 	}
 }
 
-pub async fn start_beefy_gadget<Block, Pair, Backend, Client, Network, SyncOracle>(
-	client: Arc<Client>,
+/// Start the BEEFY gadget.
+///
+/// This is a thin shim around running and awaiting a BEEFY worker. The [`Client`]
+/// convenience trait is not used here on purpose. We don't want to leak it into the
+/// public interface of the BEEFY gadget.
```

Comment:
> We don't want to leak it into the public interface of the BEEFY gadget.

Any compelling reason why we we don't want to expose it?

Reply:
Main reason, why I didn't want to expose it is that it would require to change every node which is using the BEEFY gadget. Also `Client` could be seen as a minimal bound and a client is actually free to pass a more 'rich' client.

Reply:
That's fair, although since the trait has blanket-impl, would that really be an issue? Like no one would really manually implement that trait anyway, no?

Reply:
I will turn that into an issue which should be a good one get done once things stabilize a bit more.

Reply:
See #122